pkg/sql/opt/metadata.go

```diff
@@ -56,55 +57,18 @@ type privilegeBitmap uint32
 //   -- [0] -> x
 //   -- [1] -> y
 //
-// Reusing column ids is dangerous and should be avoided in most cases. From the
-// optimizer's perspective, any column with the same id is the same column.
-// Therefore, using the same column id to represent two different columns can
-// produce inaccurate plan costs, plans that are semantically inequivalent to
-// the original plan, or errors. Columns of different types must never use the
-// same id. Additionally, column ids cannot be overloaded within two relational
-// expressions that interact with each other.
+// An operator is allowed to reuse the column ids of its input if for every
+// output row there exists an input row having identical values for those
+// columns. For example, is it safe for a Select to use its input's column ids
+// because it only filters rows. Likewise, pass-through column ids of a Project
+// can be reused.
```

Comment:
You need to add the NULL case as well. Maybe this:
```
// An operator is allowed to reuse some or all of the column ids of an input if:
//
// 1. For every output row, there exists at least one input row having identical
//    values for those columns.
// 2. OR if no such input row exists, there is at least one output row having
//    NULL values for all those columns (e.g. when outer join NULL-extends).
//
// For example, is it safe for a Select to use its input's column ids because it
// only filters rows. Likewise, pass-through column ids of a Project can be
// reused.
```

Reply:
Ahh I missed that addendum. Fixed.